src/main/java/ru/urururu/sanity/cpp/tools/Tool.java

```diff
@@ -1,28 +1,61 @@
 package ru.urururu.sanity.cpp.tools;
 
+import java.io.BufferedReader;
 import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.Collections;
+import java.util.List;
 import java.util.Optional;
 import java.util.Set;
-import java.util.function.Function;
+import java.util.function.BiFunction;
 
 /**
  * @author <a href="mailto:dmitriy.g.matveev@gmail.com">Dmitry Matveev</a>
  */
 public abstract class Tool {
+    final String executable;
+    private final List<String> versionIds;
+
+    Tool(String executable, String version) {
+        this.executable = executable;
+        this.versionIds = evaluateVersionIds(version);
+        System.out.println("version = " + version);
+        System.out.println("versionId = " + versionIds);
+    }
+
     abstract Set<Language> getLanguages();
 
-    protected static Optional<Tool> tryCreate(String executable, Function<String, Tool> factory) throws InterruptedException {
+    static Optional<Tool> tryCreate(String executable, BiFunction<String, String, Tool> factory) throws InterruptedException {
+        String version;
+
         ProcessBuilder pb = new ProcessBuilder(executable, "--version");
         try {
-            pb.start();
-            int result = pb.start().waitFor();
+            Process process = pb.start();
+
+            try (BufferedReader reader =
+                         new BufferedReader(new InputStreamReader(process.getInputStream()))){
+                version = reader.readLine();
+            }
+
+            pb.start().waitFor();
         } catch (IOException e) {
             return Optional.empty();
         }
 
-        return Optional.of(factory.apply(executable));
-
+        return Optional.of(factory.apply(executable, version));
     }
 
     public abstract String[] createParameters(String filename, String objFile);
+
+    /**
+     * @return version identifiers from most specific to more generic
+     */
+    public List<String> getVersionIds() {
+        return versionIds;
+    }
+
+    List<String> evaluateVersionIds(String version) {
+        System.err.println("unknown version = " + version);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Replace this usage of System.out or System.err by a logger. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS106)